src/StackDeploy/Wait.hs

```diff
@@ -0,0 +1,156 @@
+  , StackEvent
+  , seClientRequestToken
+  , seEventId
+  , seResourceStatus
+  , seResourceType
+  )
+import StackDeploy.AWS
+import StackDeploy.Prelude
+import StackDeploy.Types
+
+-- | Wait for remote operation to end in a final status
+--
+-- Apply action for each event related to the remote operation.
+waitForAccept :: forall m r . (AWSConstraint r m, MonadAWS m)
+              => RemoteOperation
+              -> (StackEvent -> m ())
+              -> m RemoteOperationResult
+waitForAccept remoteOperation action =
+  classify =<< runConduit (allEvents remoteOperation .| iterM action .| find (isStackEvent finalResourceStatus))
+  where
+    classify :: Maybe StackEvent -> m RemoteOperationResult
+    classify = maybe
+      (throwM $ AssertionFailed "No last event")
+      (maybe (throwM $ AssertionFailed "Last event has no resource status") remoteOperationResult . view seResourceStatus)
+
+    remoteOperationResult :: ResourceStatus -> m RemoteOperationResult
+    remoteOperationResult = \case
+      CreateComplete   -> pure RemoteOperationSuccess
+      CreateFailed     -> pure RemoteOperationFailure
+      DeleteComplete   -> pure RemoteOperationSuccess
+      DeleteFailed     -> pure RemoteOperationFailure
+      UpdateComplete   -> pure RemoteOperationSuccess
+      UpdateFailed     -> pure RemoteOperationFailure
+      RollbackComplete -> pure RemoteOperationFailure
+      RollbackFailed   -> pure RemoteOperationFailure
+      status           -> throwM . AssertionFailed $ message status
+      where
+        message :: Show a => a -> String
+        message status = "Last event has unexpected resource status: " <> show status
+
+-- | Conduit polling for new stack events related to a remote operation
+allEvents :: forall m r . (AWSConstraint r m, MonadAWS m)
+          => RemoteOperation
+          -> ConduitT () StackEvent m ()
+allEvents remoteOperation = go =<< getLastEvent =<< getInitial
+  where
+    getLastEvent :: [StackEvent] -> ConduitT () StackEvent m StackEvent
+    getLastEvent = maybe (throwM $ AssertionFailed "No initial event") pure . listToMaybe
+
+    initialEvents :: ConduitT () StackEvent m ()
+    initialEvents
+      = stackEvents remoteOperation
+      $ takeUntilInclusive (isStackEvent initialResourceStatus)
+
+    nextEvents :: StackEvent -> ConduitT () StackEvent m ()
+    nextEvents lastEvent
+      = stackEvents remoteOperation
+      $ takeWhile (not . isExpectedEvent lastEvent)
+
+    getInitial = do
+      events <- poll initialEvents
+      if null events
+        then delay >> getInitial
+        else pure events
+
+    go :: StackEvent -> ConduitT () StackEvent m ()
+    go lastEvent = do
+      delay
+      events <- poll (nextEvents lastEvent)
+      if null events
+        then go lastEvent
+        else maybe (pure ()) go (listToMaybe events)
+
+    isExpectedEvent :: StackEvent -> StackEvent -> Bool
+    isExpectedEvent left right = view seEventId left == view seEventId right
```

Comment:
Using `Data.Function.on` this can be simplified to:

```haskell
isExpectedEvent = (==) `on` view seEventId
```

Reply:
good one I forgot about that option.